modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/db/Migration.scala

```diff
@@ -181,54 +180,38 @@ object Migration {
   def empty[F[_]: Applicative]: Migration[F] =
     Migration[F](Nil, Applicative[F].unit)
 
-  def buildBlock[F[_]: MonadThrow: DAO, I](description: Description, target: Target[I]): F[Option[Block]] =
+  def buildBlock[F[_]: MonadThrow: DAO, I](description: Description, target: Target[I]): F[List[Block]] =
     description match {
       case Description.Table(schemas) =>
-        val tableName = StringUtils.getTableName(schemas.latest)
-
-        val migrate: F[Option[Block]] = for {
-          schemaKey <- getVersion[F](tableName)
-          matches = schemas.latest.schemaKey == schemaKey
-          block <- if (matches) emptyBlock[F]
-                   else
-                     Control.getColumns[F](tableName).flatMap { (columns: List[ColumnName]) =>
-                       migrateTable(target, schemaKey, columns, schemas) match {
-                         case Left(migrationError) => MonadThrow[F].raiseError[Option[Block]](migrationError)
-                         case Right(block) => MonadThrow[F].pure(block.some)
-                       }
-                     }
-        } yield block
-
-        Control.tableExists[F](tableName).ifM(migrate, Monad[F].pure(target.createTable(schemas).some))
+        val schemaKeysWithModels = foldMapMergeRedshiftSchemas(schemas)
+        val goodModel = schemaKeysWithModels.goodModel
+        val goodTableName = goodModel.tableName
+        val highestSchemaKey = schemas.last.self.schemaKey
```

Comment:
There are two different things we might mean when we say "highest schema key":

1. The highest schema key that exists in the Iglu Server
2. The highest schema key that was present in the batch of events.

Here I believe you are using the first definition.  As a consequence, whenever somebody publishes a new version of a schema, RDB loader will **immediately** migrate the table to the newest version.

To be fair, RDB loader has already been like that.  But our experience tells us it's not helpful to have the loader over-eager to migrate tables.  It is much more helpful if the loader only migrates the table when the user starts tracking events with the new schema.  i.e. it's better to use the 2nd definition, and to only migrate the tables as far as the highest actively used schema version.

You might think I am moving the goalposts (sorry!), but I was really hoping the new implementation would fix the old problem of over-eager migrations.

Reply:
thanks @istreeter ! I agree and will update PR